services/src/main/java/io/grpc/services/HealthServiceImpl.java

```diff
@@ -103,22 +110,48 @@ public void cancelled(Context context) {
 
   void setStatus(String service, ServingStatus status) {
     synchronized (watchLock) {
-      ServingStatus prevStatus = statusMap.put(service, status);
-      if (prevStatus != status) {
-        notifyWatchers(service, status);
+      if (terminal) {
+        logger.log(Level.FINE, "Ignoring status {} for {}", new Object[]{status, service});
+        return;
       }
+      setStatusInternal(service, status);
+    }
+  }
+
+  @GuardedBy("watchLock")
+  private void setStatusInternal(String service, ServingStatus status) {
+    ServingStatus prevStatus = statusMap.put(service, status);
+    if (prevStatus != status) {
+      notifyWatchers(service, status);
     }
   }
 
   void clearStatus(String service) {
     synchronized (watchLock) {
+      if (terminal) {
+        logger.log(Level.FINE, "Ignoring status clearing for {}", new Object[]{service});
+        return;
+      }
       ServingStatus prevStatus = statusMap.remove(service);
       if (prevStatus != null) {
         notifyWatchers(service, null);
       }
     }
   }
 
+  void enterTerminalState() {
+    synchronized (watchLock) {
+      if (terminal) {
+        logger.warning("Already terminating");
```

Comment:
If you consider this a bug, you should probably log a stack trace here to help debugging.
If it's not a bug, a warning wouldn't be necessary.

Reply:
Done.